### PR TITLE
Add avoid_breaking_exported_api config option

### LINKED ISSUE #4504: Option to disable trivially_copy_pass_by_ref in public functions
Running Clippy over Rust repo yields about 100 unique warnings from `trivially_copy_pass_by_ref`, almost all of them were tripped by public functions which are not going to change.

Having option to disable it for public functions would help if this and probably other cases.

EDIT: `ptr_arg` have similar issue (but there are only 13 occurrences). Maybe something more generic should be developed?

### LINKED ISSUE #6806: Config for public API changes
Sometimes we turn off lints in cases where the suggestion would change the public API of a crate. This is a reasonable default, but I believe this should be an opt-out configuration option for all Clippy lints. Oftentimes, a crate has a public API but it is still "private" in the sense that it is only used within a single project (e.g. `clippy_utils`), and so API changes are not as much of a concern.

### PR DESCRIPTION
**Pull request by camsteffen**
changelog: Add `avoid_breaking_exported_api` config option for [`enum_variant_names`], [`large_types_passed_by_value`], [`trivially_copy_pass_by_ref`], [`unnecessary_wraps`], [`upper_case_acronyms`] and [`wrong_self_convention`].

changelog: Deprecates [`pub_enum_variant_names`] and [`wrong_pub_self_convention`] as the non-pub variants are now configurable.

changelog: Fix various false negatives for `pub` items that are not exported from the crate.

A couple changes to late passes in order to use `cx.access_levels.is_exported` rather than `item.vis.kind.is_pub`.

I'm not sure how to better document the config option or lints that are (not) affected (see comments in #6806). Suggestions are welcome. cc @rust-lang/clippy

I added `/clippy.toml` to use the config internally and `/tests/clippy.toml` to maintain a default config in ui tests.

Closes #6806
Closes #4504

### CODE REVIEW

File: clippy.toml [hunk 1]
```diff
@@ -0,0 +1 @@
+avoid-breaking-exported-api = false
```
Comment (flip1995) [inline clippy.toml:1]: Now we went full circle with dogfooding Clippy! :smile:

File: tests/clippy.toml [hunk 2]
```diff
@@ -0,0 +1 @@
+# default config for tests
```
Comment (flip1995) [inline tests/clippy.toml:1]: Please extend this comment to explain that this file is necessary to override the root `clippy.toml` file.

File: clippy_lints/src/deprecated_lints.rs [hunk 3]
```diff
@@ -134,3 +134,22 @@ declare_deprecated_lint! {
...
     "this lint has been replaced by `manual_filter_map`, a more specific lint"
 }
+
+declare_deprecated_lint! {
+    /// **What it does:** Nothing. This lint has been deprecated.
+    ///
+    /// **Deprecation reason:** The `avoid_breaking_exported_api` config option was added, which
+    /// enables the `enum_variant_names` lint for public items.
+    /// ```
+    pub PUB_ENUM_VARIANT_NAMES,
```
Comment (flip1995) [inline clippy_lints/src/deprecated_lints.rs:144]: Please also add tests for this to `tests/ui/deprecated.rs`.

**Review by flip1995 — COMMENTED** [review]
Haven't reviewed it in-depth (just reviewed the "infra"-impact), but this looks like a great improvement of Clippy. Thanks!

File: clippy_lints/src/lib.rs [hunk 4]
```diff
@@ -503,6 +492,14 @@ pub fn register_plugins(store: &mut rustc_lint::LintStore, sess: &Session, conf:
         "clippy::filter_map",
         "this lint has been replaced by `manual_filter_map`, a more specific lint",
     );
+    store.register_removed(
+        "clippy::pub_enum_variant_names",
+        "use the `avoid_breaking_exported_api` config option to enable the `enum_variant_names` lint for public items",
```
Comment (phansch) [inline clippy_lints/src/lib.rs:497]: A bit more straightforward:

```suggestion
        "set the `avoid_breaking_exported_api` config option to `false` to enable the `enum_variant_names` lint for public items",
```

**Review by phansch — APPROVED** [review]
Thanks for the PR! The implementation looks good to me. My only concern is that we don't have tests for when the option is set to false. I think this might be a general problem though, and I wouldn't object to merging the PR anyway. What do you think?

**Comment by camsteffen** [comment]
Rebased with your comment.

You're right that `false` is not tested. Though it is proven out with the dogfood test. I'm okay with it if you are :)

**Comment by phansch** [comment]
Thanks!
@bors r=flip1995,phansch

**Pull request merged**